### ISSUE TITLE
web throw NullPointerException: hostname must not be null

### ISSUE THREAD

**Issue by aiqing2171** [naver/pinpoint#3985]
**Pinpoint version**:1.8.0-SNAPSHOT  today's newest.

If i set inbound=1, it will not throw exception(receive message from rabbit mq). But if i set inbound=2, pinpoint web shows "Error!  hostname must not be null" 

@jiaqifeng  Thank you for looking at this!

```
2018-04-04 17:30:54 [WARN ](.n.p.w.c.ControllerExceptionHandler) Failed to execute controller methods. message:java.lang.NullPointerException: hostname must not be null, request:{method=GET, heads={referer=[http://***:8080/], accept-language=[zh-CN,zh;q=0.9], cookie=[_gat=1; _ga=GA1.1.893457360.1519869019; _gid=GA1.1.1073140571.1522632488], host=[***:8080], x-requested-with=[XMLHttpRequest], connection=[keep-alive], accept-encoding=[gzip, deflate], accept=[application/json, text/javascript, */*; q=0.01], user-agent=[Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.62 Safari/537.36]}, parameters={wasOnly=[false], bidirectional=[false], callerRange=[2], from=[1522833947000], to=[1522834247000], calleeRange=[2], serviceTypeName=[TOMCAT], applicationName=[*_bps], _=[1522834248602]}, url=http://***:8080/getServerMapDataV2.pinpoint}.
java.util.concurrent.CompletionException: java.lang.NullPointerException: hostname must not be null
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1592)
        at om.navercorp.pinpoint.web.task.RequestContextPropagatingTaskDecorator$RequestContextPropagatingRunnable.run(RequestContextPropagatingTaskDecorator.java:50)
        at com.navercorp.pinpoint.web.task.SecurityContextPropagatingTaskDecorator$SecurityContextPropagatingRunnable.run(SecurityContextPropagatingTaskDecorator.java:50)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: hostname must not be null
        at com.navercorp.pinpoint.web.applicationmap.rawdata.LinkData.addLinkData(LinkData.java:69)
        at com.navercorp.pinpoint.web.applicationmap.rawdata.LinkDataMap.addLinkData(LinkDataMap.java:44)
        at com.navercorp.pinpoint.web.mapper.MapStatisticsCalleeMapper.mapRow(MapStatisticsCalleeMapper.java:106)
        at com.navercorp.pinpoint.web.mapper.MapStatisticsCalleeMapper.mapRow(MapStatisticsCalleeMapper.java:45)
        at com.navercorp.pinpoint.web.mapper.RowMapReduceResultExtractor.extractData(RowMapReduceResultExtractor.java:26)
        at com.navercorp.pinpoint.common.hbase.HbaseTemplate2.executeParallelDistributedScan(HbaseTemplate2.java:676)
        at com.navercorp.pinpoint.common.hbase.HbaseTemplate2.findParallel(HbaseTemplate2.java:658)
        at com.navercorp.pinpoint.web.dao.hbase.HbaseMapStatisticsCalleeDao.selectCallee(HbaseMapStatisticsCalleeDao.java:96)
        at com.navercorp.pinpoint.web.service.LinkDataMapServiceImpl.selectCalleeLinkDataMap(LinkDataMapServiceImpl.java:46)
        at com.navercorp.pinpoint.web.service.map.DefaultApplicationMapCreator.createMap(DefaultApplicationMapCreator.java:84)
        at com.navercorp.pinpoint.web.service.map.DefaultApplicationsMapCreator$1.get(DefaultApplicationsMapCreator.java:90)
        at com.navercorp.pinpoint.web.service.map.DefaultApplicationsMapCreator$1.get(DefaultApplicationsMapCreator.java:87)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        ... 5 more
```

**Comment by Xylus** [comment]
Hi @aiqing2171 
Thanks for reporting.
This might be due to a bug in RabbitMQ plugin, we'll take a look into this asap.

Could you also describe how your RabbitMQ producers and clients are set up in as much detail as possible please?
Things like how your message production and consumption are set up (ie on top of spring-amqp-rabbit? if so, how your listeners are registered etc, if not, how your producers/consumers are set up etc) would help a great deal in narrowing down the cause.

Thanks.

**Comment by Xylus** [comment]
@aiqing2171 
We couldn't get it reproduced.
Could you tell us your rabbitmq client version? (or spring-amqp-rabbit version if you're using this).
DEBUG level logs of your collector, web and the agent that caused the issue would also be of great help.

**Comment by aiqing2171** [comment]
Hi @Xylus 

amqp-client.version:3.5.1

Now pinpoint web shows  ''rabbitmq  exchange *** ->my tomcat application'.
In fact, my application sends message to this queue and receives message from it. 
But it only shows receiving mesage from mq. 
This problem may have something with the way we use mq api.  We just called the api.
This below is the code:

> final RabbitTemplate rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMandatory(true);
        rabbitTemplate.setExchange(exchange);
        rabbitTemplate.setRoutingKey(routingKey);
        rabbitTemplate.setMessageConverter(new Jackson2JsonMessageConverter());
        rabbitTemplate.setConfirmCallback(new RabbitTemplate.ConfirmCallback() {
        rabbitTemplate.convertAndSend(message);
....


For message consumer:

> executor = Executors.newFixedThreadPool(threadPoolConsumerBuilder.threadCount);
executor.execute(new Runnable() {
        @Override
         public void run() {
          while (!stop) {
           try {                           
            DetailRes detailRes = messageConsumer.consume();
             ......
} ...
The core code in messageConsumer class:
Channel channel = connection.createChannel(false);
 QueueingConsumer consumer = new QueueingConsumer(channel);
channel.basicQos(0, 1, false);
channel.basicConsume(queue, false, consumer);
...
Channel channel = consumer.getChannel();
delivery = consumer.nextDelivery();
Message message = new Message(delivery.getBody(), messagePropertiesConverter
                            .toMessageProperties(delivery.getProperties(), delivery.getEnvelope(), "UTF-8"));

**Comment by aiqing2171** [comment]
Found another problem in pinpoint collector.

>2018-04-09 10:56:55 [WARN ](c.n.p.c.h.SpanHandler              ) Span handle error. Caused:calleeApplicationName must not be null. 
Span:TSpan(agentId:224-agent, applicationName:app1-agent, agentStartTime:1522813069650, 
transactionId:00 1C 32 32 2D 73 6D 73 2D 72 65 63 65 69 76 65 EC CB B6 BC A7 2C AD D0 07,
 spanId:-5165118704577994137, parentSpanId:-384252399272208572, 
 startTime:1523242615425, elapsed:7, rpc:rabbitmq://exchange=Ex.com.e.topic, 
 serviceType:8300, endPoint:exchange:Ex.com.e.topic, remoteAddr:*.222:5672, 
 annotations:[TAnnotation(key:131, value:<TAnnotationValue stringValue:Kcom.*.handle.queue1>)], 
 flag:0, err:1, spanEventList:[TSpanEvent(sequence:2, startElapsed:1, endElapsed:1, serviceType:2301, endPoint:*.11:1521, 
 annotations:[TAnnotation(key:20, value:<TAnnotationValue intStringStringValue:TIntStringStringValue(intValue:4, stringValue1:0,1, stringValue2:1, S04)>)], 
 depth:3, destinationId:oracledb, apiId:-28), TSpanEvent(sequence:1, startElapsed:0, endElapsed:2, 
 serviceType:5510, annotations:[TAnnotation(key:-30, value:<TAnnotationValue intValue:-3>)], depth:2, apiId:-18), 
 TSpanEvent(sequence:0, startElapsed:0, endElapsed:2, serviceType:5071, depth:1, apiId:-17), 
 TSpanEvent(sequence:3, startElapsed:2, endElapsed:2, serviceType:9110, endPoint:*.143:20883, 
 annotations:[TAnnotation(key:90, value:<TAnnotationValue stringValue:[Ljava.lang.Object;@5a2263ed>), 
 TAnnotation(key:91, value:<TAnnotationValue stringValue:RpcResult [result=-1, exception=null]>)], 
 depth:1, nextSpanId:-6063426377638391324, destinationId:*.143:20883, apiId:-38), 
 TSpanEvent(sequence:4, startElapsed:4, serviceType:9110, nextSpanId:-2585515602390443052, apiId:-38, 
 exceptionInfo:TIntStringValue(intValue:3, stringValue:Failed to invoke the method sendMt in the service com.*.RequestService.
 No provider available for the service com.*.RequestService from registry *.143:2181 on the consumer *.224 using th...(349))), 
 TSpanEvent(sequence:7, startElapsed:5, endElapsed:1, serviceType:2301, endPoint:*.11:1521, 
 annotations:[TAnnotation(key:20, value:<TAnnotationValue intStringStringValue:TIntStringStringValue(intValue:-7, stringValue2:-2, 1112755)>)],
 depth:3, destinationId:testdb, apiId:-28), TSpanEvent(sequence:6, startElapsed:5, endElapsed:2, serviceType:5510, 
 annotations:[TAnnotation(key:-30, value:<TAnnotationValue intValue:5>)], depth:2, apiId:24), 
 TSpanEvent(sequence:5, startElapsed:5, endElapsed:2, serviceType:5071, apiId:8)], 
 parentApplicationName:app1-agent, parentApplicationType:1010, acceptorHost:exchange-Ex.com.e.topic, apiId:-4, applicationServiceType:1010, loggingTransactionInfo:0)
java.lang.NullPointerException: calleeApplicationName must not be null
        at com.navercorp.pinpoint.collector.dao.hbase.HbaseMapStatisticsCallerDao.update(HbaseMapStatisticsCallerDao.java:91)
        at com.navercorp.pinpoint.collector.handler.StatisticsHandler.updateCaller(StatisticsHandler.java:59)
        at com.navercorp.pinpoint.collector.handler.SpanHandler.insertSpanEventStat(SpanHandler.java:181)
        at com.navercorp.pinpoint.collector.handler.SpanHandler.handleSimple(SpanHandler.java:88)
        at com.navercorp.pinpoint.collector.receiver.AbstractDispatchHandler.dispatchSendMessage(AbstractDispatchHandler.java:58)
        at com.navercorp.pinpoint.collector.receiver.DispatchHandlerWrapper.dispatchSendMessage(DispatchHandlerWrapper.java:47)
		at com.navercorp.pinpoint.collector.receiver.udp.BaseUDPHandlerFactory$DispatchPacket.receive(BaseUDPHandlerFactory.java:89)
        at com.navercorp.pinpoint.collector.receiver.udp.BaseUDPHandlerFactory$DispatchPacket.receive(BaseUDPHandlerFactory.java:68)
        at com.navercorp.pinpoint.collector.receiver.udp.Task.run(Task.java:44)
        at com.navercorp.pinpoint.collector.monitor.MonitoredRunnableDecorator$InstrumentedRunnable.run(MonitoredRunnableDecorator.java:113)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

**Comment by Xylus** [comment]
@aiqing2171 
Thanks for the detailed log, it'll help a lot in figuring out the issue.
I'll take a look into it and let you know of any progress.
Thanks again!

**Comment by Xylus** [comment]
@aiqing2171 
So I think there are 2 issues here (please correct me if I'm mistaken) :
1. Message producer not being traced
2. *java.lang.NullPointerException: calleeApplicationName must not be null* thrown in collector.

The second issue I think is caused by Dubbo plugin when an exception is thrown (`Failed to invoke the method sendMt in the service ...` in the trace log above). **DubboConsumerInterceptor** records `destinationId` only when no exception is thrown, so this will lead to NPE like above when the span event is handled by the collector. This should be fixed in a separate issue, so let's leave this out for now in this issue.

The first issue I think is definitely relevant to RabbitMQ plugin.
Could you explain how your producer is run with a bit more detail? Any little bit of information will help such as:
Is your producer being traced at all?
Is the agent on your producer producing any error log? or any log at all?
Is the **RabbitTemplate** code above executed in a separate thread? etc

Thanks.

**Comment by aiqing2171** [comment]
@Xylus 
This is a little complex. There are two kinds of mq message producer.
1. rabbitmq exchange-*-ack, as the pinpoint web shows, the producer（app2）which put message to ack queue can be traced and viewed in pinpoint call tree.
2. my tomcat application produces messages by itself and consumes this message(in order to retry,it rarely happened). It relates with two queues，but they are not displayed in web server map.

Nothing wrong found in the agent log.
RabbitTemplate code above executed in a separate thread.

'app2'  is a quartz crontab spring project deployed in tomcat. I set the crontab job class for traced in agent.config entrypoint.  But,it also throw exception 'hostname must not be null' when selecting application 'app2'  and inbound = 2.

**Comment by Xylus** [comment]
@aiqing2171 
Thanks for the explanation.

To be clear, it's your tomcat application that executes the **RabbitTemplate** code in a separate thread, right? (The second one from your example).
If so, is everything up to the thread switch is traced correctly?

*app2* (the quartz crontab spring project) executes the **RabbitTemplate** code in the same thread (as the one executing the entry point method) and is traced okay?
(except for the fact that inbound = 2 queries will throw *hostname must not be null* exception)

**Comment by aiqing2171** [comment]
Perhaps this below is more clear.

a. 'app2' quartz spring project deployed in tomcat ->ack queue/ack queue -> my tomcat application.
app2 executes the RabbitTemplate code in the same thread.
b. dubbo consumer ->my tomcat application->（sometimes my tomcat application produces message to ) warn queue/ack queue->my tomcat application.   
"_my tomcat application-> warn queue/ack queue_" can not find any trace info in call tree for this step.

**Comment by aiqing2171** [comment]
should RabbitMq application create an 'agentId' when putting into hbase?

**Comment by Xylus** [comment]
@aiqing2171 Yes, but it should be set automatically via the plugin.
I'm trying to recreate your app2 environment to see if I can reproduce your issue.

Since you're on the right track, another part you could check is collector's **SpanHandler**.
This is where the data used in `com.navercorp.pinpoint.web.mapper.MapStatisticsCalleeMapper` gets populated.
For *callerHost* to be null, the 5th argument of `StatisticsHandler.updateCallee(...)` should be null or empty, which shouldn't happen as they simply use `span.getAgentId()`.
It'd be a good idea to debug and check what gets passed here (this is the part where I'm trying to figure out what's going on as well)

**Comment by Xylus** [comment]
By the way, this has mostly to do with case a) 'app2' quartz spring project deployed in tomcat ->ack queue/ack queue -> my tomcat application.

For the b) case where the tomcat application occasionally calls the warn/ack queue in another thread, I don't think the RabbitMQ invocation is traced as they are executed in a different thread.

**Comment by aiqing2171** [comment]
RabbitMQClientPlugin.java
```
transformTemplate.transform("org.springframework.amqp.rabbit.core.RabbitTemplate", new TransformCallback() {
   @Override
   public byte[] doInTransform(Instrumentor instrumentor, ClassLoader loader, String className, 
Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws InstrumentException {
    InstrumentClass target = instrumentor.getInstrumentClass(loader, className, classfileBuffer);
    for (InstrumentMethod publicApi : target.getDeclaredMethods(publicApiFilter)) {
		publicApi.addScopedInterceptor(BasicMethodInterceptor.class.getName(), va(RabbitMQClientConstants.RABBITMQ_CLIENT_INTERNAL), RabbitMQClientConstants.RABBITMQ_TEMPLATE_API_SCOPE);
    }
    InstrumentMethod invoke = target.getDeclaredMethod("invoke", "org.springframework.amqp.rabbit.core.RabbitOperations$OperationsCallback");
		if (invoke != null) {
            invoke.addInterceptor(BasicMethodInterceptor.class.getName(), va(RabbitMQClientConstants.RABBITMQ_CLIENT_INTERNAL));
        }
```
BasicMethodInterceptor do not set agentId for trace?

**Comment by Xylus** [comment]
I've managed to recreate the issue.
When the server map is rendered, virtual queue node uses the end point recorded in the span to emulate upstream nodes.
However, when a custom entry point is set for the producing node, the span created via **UserIncludeMethodInterceptor** does not record an end point because... there is no end point for user created spans.

This issue should be reproducible under the following situation:
1) Message producer starts a trace using a **user-defined entry point**.
2) Message consumer operates normally, consuming messages from the producer and tracing everything.
3) Query the server map with the consumer selected, and set inbound to > 2 in order to emulate a link between the producer node and the virtual queue node.

I'll look into this a bit more and try to come up with a fix.
Thanks.

**Comment by aiqing2171** [comment]
Thanks!  *_*

**Comment by jiaqifeng** [comment]
@abhijith0505 
From your exception stacktrace and span content I could not find any clue about rabbitmq. Only about the dubbo do not record destinationId bug. I prefer you fix the dubbo plugin and try again.

**Comment by Xylus** [comment]
Merged for 1.7.3
